File: src/main/java/io/github/hiiragi283/api/HTMaterialsAPI.java
package io.github.hiiragi283.api;

import net.minecraft.util.ResourceLocation;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import io.github.hiiragi283.HMReference;
import io.github.hiiragi283.api.extension.ServiceLoaderUtil;
import io.github.hiiragi283.api.material.HTMaterialRegistry;
import io.github.hiiragi283.api.part.HTPartManager;
import io.github.hiiragi283.api.shape.HTShapeRegistry;

public interface HTMaterialsAPI {

    String MOD_ID = HMReference.MOD_ID;
    String MOD_NAME = HMReference.MOD_NAME;
    String VERSION = HMReference.VERSION;
    Logger LOGGER = LogManager.getLogger(MOD_NAME);
    HTMaterialsAPI INSTANCE = ServiceLoaderUtil.getSingleton(HTMaterialsAPI.class);

    static ResourceLocation location(String path) {
        return new ResourceLocation(MOD_ID, path);
    }

    HTShapeRegistry shapeRegistry();

    HTMaterialRegistry materialRegistry();

    HTPartManager partManager();
}

File: src/main/kotlin/io/github/hiiragi283/api/HTMaterialsAPI.kt
package io.github.hiiragi283.api

import io.github.hiiragi283.api.extension.runCatchAndLog
import io.github.hiiragi283.api.item.HTMaterialItemProvider
import io.github.hiiragi283.api.material.HTMaterialRegistry
import io.github.hiiragi283.api.part.HTPart
import io.github.hiiragi283.api.part.HTPartManager
import io.github.hiiragi283.api.shape.HTShapeKey
import io.github.hiiragi283.api.shape.HTShapeRegistry
import io.github.hiiragi283.common.core.HTMaterialsAPIImpl
import net.minecraft.util.ResourceLocation

interface HTMaterialsAPI {

    companion object {
        const val MOD_ID: String = "ht_materials"
        const val MOD_NAME: String = "HT Materials Legacy"
        const val MOD_VERSION: String = "0.0.1"

        @JvmStatic
        fun id(path: String): ResourceLocation = ResourceLocation(MOD_ID, path)

        @JvmStatic
        val instance: HTMaterialsAPI = HTMaterialsAPIImpl
    }

    val materialRegistry: HTMaterialRegistry
    val shapeRegistry: HTShapeRegistry
    val partManager: HTPartManager

    val materialItemMap: Map<HTShapeKey, HTMaterialItemProvider>

    /*fun forEachPhasedMaterial(action: (HTPhasedMaterial) -> Unit) {
        materialRegistry.keys.forEach { materialKey ->
            HTFluidPhase.entries.map { phase ->
                HTPhasedMaterial(materialKey, phase)
            }.forEach { runCatchAndLog { action(it) } }
        }
    }*/

    fun forEachPart(action: (HTPart) -> Unit) {
        materialRegistry.keys.forEach { materialKey ->
            shapeRegistry.keys.map { shapeKey ->
                HTPart(materialKey, shapeKey)
            }.forEach { runCatchAndLog { action(it) } }
        }
    }

    fun forEachAddon(action: (HTMaterialsPlugin) -> Unit)

}